src/handle/command.ml

```diff
@@ -407,38 +405,90 @@ let handle : (Path.t -> Sign.t) -> sig_state -> p_command ->
             (* Just ignore the command with a warning. *)
             wrn pe.pos "Proof aborted."; ss
         | _ ->
-            (* Check that no metavariable remains. *)
-            if LibTerm.Meta.has true a then
-              (fatal_msg "The type of %a has unsolved metavariables.\n"
-                 pp_uid id;
-               fatal pe.pos "We have %a : %a." pp_uid id pp_term a);
-            (match t with
-             | Some(t) when LibTerm.Meta.has true t ->
-                 fatal_msg
-                   "The definition of %a has unsolved metavariables.\n"
-                   pp_uid id;
-                 fatal pe.pos "We have %a : %a ≔ %a."
-                   pp_uid id pp_term a pp_term t
-             | _ -> ());
             match pe.elt with
             | P_proof_abort -> assert false (* Handled above *)
             | P_proof_admit ->
                 (* If the proof is finished, display a warning. *)
                 if finished ps then
                   wrn pe.pos "The proof is finished. Use 'end' instead.";
+                (* Get a list of metas, ordered by dependency. *)
+                let metas =
+                  let ms =
+                    let f = function Typ g -> Some g.goal_meta | _ -> None in
+                    List.filter_map f ps.proof_goals
+                  in
+                  let dep m_a m_b =
+                    if m_a == m_b then 0 else
+                    if LibTerm.Meta.(m_a < m_b) then -1 else
+                    if LibTerm.Meta.(m_b < m_a) then 1 else 0
+                    (* Incomparable meta variables are considered equal. *)
```

Comment:
This looks wrong. Imagine you have a2 > a1 and b incomparable with both a1 and a2. Then, dep a2 b = dep b a1 = 0. So, by transitivity, we should have a2 = a1...
The problem is that your dependency order can only be partial. So you can sort metavariables with such an ordering.
We therefore need another strategy.
Why not simply iterating other metavariables, recording those already seen, and raise an exception when we encounter a metavariable already seen?
```
exception Cannot_admit
(* ms is the set of metas already seen *)
let rec admit_meta ms m =
  if MetaSet.mem m ms then raise Cannot_admit;
  iter true (admit ms) !(m.meta_type);
  ... (* code adding a new symbol f of type !(m.meta_type) and setting the value of m to f *)
in
let admit_goal g =
  match g with
  | Unif _ -> ()
  | Typ gt ->
  let m = gt.goal_meta in
  match !(m.meta_value) with
  | Some _ -> ()
  | None -> admit_meta MetaSet.empty gt.goal_meta
in
List.iter admit_goal ps.proof_goals
```

Reply:
I implemented something based on this idea. We should note that we shouldn't obtain the `Cannot_admit` because the typechecker should produce admissible terms.